### DISCUSSION TITLE
Testing Observations

### DISCUSSION THREAD

**Discussion by dsilhavy** [5G-MAG/rt-5gms-application-function#97] [General]
# Description
I am using this discussion thread to share some of my findings while testing different functionality of the Application Function. The idea is for other people to comment and share their findings before creating an issue. Potentially we can also convert this discussion to an issue at some point.

## Installation based on main branch
While the installation of the Application Function worked fine for the `development` branch with the latest commit (https://github.com/5G-MAG/rt-5gms-application-function/commit/d5af8a6e64e8db1e53f244f45b70566d183e48a7) I ran into an issue when installing based on the `main` branch with commit (https://github.com/5G-MAG/rt-5gms-application-function/commit/7cbc6fbf2332a8f25bf83f9deb82f2f82e120d29). In both cases a clean installation from scratch was performed including a `git clone` and strictly following the installation guidelines as described in the Readme.

The following error occurs when installing from the `main` branch and trying to run the Application Function with `/usr/local/bin/open5gs-msafd`:

````
dsi@dsi-ThinkPad-X240:~/5gmag/rt-5gms-application-function/build$ /usr/local/bin/open5gs-msafd
/usr/local/bin/open5gs-msafd: symbol lookup error: /usr/local/bin/open5gs-msafd: undefined symbol: ogs_sbi_subscription_data_build_default

````

## M1 CRUD operations for Consumption Reporting Configuration 
Observations are based on the `development` branch with the latest commit (https://github.com/5G-MAG/rt-5gms-application-function/commit/d5af8a6e64e8db1e53f244f45b70566d183e48a7) 

### Create
Adding a consumption reporting configuration using the `m1session` tool as documented in https://github.com/5G-MAG/rt-5gms-application-function/pull/92 worked fine:

![image](https://github.com/5G-MAG/rt-5gms-application-function/assets/2427039/e5ef6119-2c1b-436d-892b-fccf24bf1cd7)

### Update

Worked fine
![image](https://github.com/5G-MAG/rt-5gms-application-function/assets/2427039/8ff2d9bd-302f-44ac-8dca-03fcd17e85cf)

#### Discussion Item 1
When updating a configuration and providing an invalid configuration shall the old values be preserved:

![image](https://github.com/5G-MAG/rt-5gms-application-function/assets/2427039/0347220e-c8b6-4076-83a1-0ff2f3033edf)

#### Discussion Item 2
It looks like we can not recover from an invalid state even if we provide valid values again. The solution here is to delete the consumption reporting configuration and provide a new valid configuration:

![image](https://github.com/5G-MAG/rt-5gms-application-function/assets/2427039/26d2e03d-1c6f-4c5f-9ec6-b1d2d980ff3a)

### Delete
Worked fine

![image](https://github.com/5G-MAG/rt-5gms-application-function/assets/2427039/1f208d57-2536-4c22-8b3f-8f14216ee1f9)

## Creating a content hosting configuration
Observations are based on the `development` branch with the latest commit (https://github.com/5G-MAG/rt-5gms-application-function/commit/d5af8a6e64e8db1e53f244f45b70566d183e48a7) 

Running the following command based on https://github.com/5G-MAG/rt-5gms-application-function/wiki/Testing-the-M1-Interface-on-v1.3.0 leads to an error and a crash of the Application Function

`./m1-session new-stream -e MyAppId -a MyASPId -n 'Big Buck Bunny' 'http://amssamples.streaming.mediaservices.windows.net/622b189f-ec39-43f2-93a2-201ac4e31ce1/BigBuckBunny.ism/' 'manifest(format=mpd-time-csf)' 'manifest(format=m3u8-aapl-v3)' `

Error:
````
10/06 08:53:47.057: [sbi] FATAL: connection_add: Assertion `ogs_app()->sbi.client.key' failed. (../subprojects/open5gs/lib/sbi/client.c:388)
10/06 08:53:47.057: [core] FATAL: backtrace() returned 14 addresses (../subprojects/open5gs/lib/core/ogs-abort.c:37)
/usr/local/lib/x86_64-linux-gnu/libogssbih1.so.2(+0x371d2) [0x7f2d353021d2]
/usr/local/lib/x86_64-linux-gnu/libogssbih1.so.2(ogs_sbi_client_send_request+0x272) [0x7f2d35303beb]
/usr/local/bin/open5gs-msafd(+0x1f094) [0x5586c3f26094]
/usr/local/bin/open5gs-msafd(+0x1dfbe) [0x5586c3f24fbe]
/usr/local/bin/open5gs-msafd(+0x1d7f5) [0x5586c3f247f5]
/usr/local/bin/open5gs-msafd(+0x27875) [0x5586c3f2e875]
/usr/local/lib/x86_64-linux-gnu/libogscore.so.2(ogs_fsm_dispatch+0x119) [0x7f2d352aadde]
/usr/local/bin/open5gs-msafd(+0x87cf) [0x5586c3f0f7cf]
/usr/local/lib/x86_64-linux-gnu/libogscore.so.2(ogs_fsm_dispatch+0x119) [0x7f2d352aadde]
/usr/local/bin/open5gs-msafd(+0x794d) [0x5586c3f0e94d]
/usr/local/lib/x86_64-linux-gnu/libogscore.so.2(+0x11a91) [0x7f2d3529ba91]
/lib/x86_64-linux-gnu/libc.so.6(+0x94ac3) [0x7f2d34494ac3]
/lib/x86_64-linux-gnu/libc.so.6(+0x126a40) [0x7f2d34526a40]
Aborted
````

This might be related to a misconfiguration on my side with the Application Server. The Application Server started in the background with the following logs:

````
dsi@dsi-ThinkPad-X240:~/5gmag/rt-5gms-application-server$ sudo PATH="usr/local/openresty/nginx/sbin:$PATH" 5gms-application-server
[sudo] password for dsi: 
[2023-10-06 09:44:22 +0200] [87194] [INFO] Running on http://127.0.0.1:7777 (CTRL + C to quit)
INFO:hypercorn.error:Running on http://127.0.0.1:7777 (CTRL + C to quit)
````

**Comment by dsilhavy** [comment]
# M5 Service Access Information - Consumption Reporting Configuration

Observations are based on the development branch with the latest commit (https://github.com/5G-MAG/rt-5gms-application-function/commit/d5af8a6e64e8db1e53f244f45b70566d183e48a7)

The configuration is included as expected

URL: http://127.0.0.24:7777/3gpp-m5/v2/service-access-information/33a4aaa0-6429-41ee-b472-0ba8822ffea4

````json
{
	"provisioningSessionId":	"33a4aaa0-6429-41ee-b472-0ba8822ffea4",
	"provisioningSessionType":	"DOWNLINK",
	"streamingAccess":	{
	},
	"clientConsumptionReportingConfiguration":	{
		"reportingInterval":	60,
		"serverAddresses":	["http://127.0.0.24:7777/3gpp-m5/v2/"],
		"locationReporting":	false,
		"samplePercentage":	100
	}
}
````

## Discussion Item:
According to TS26.512 Table 11.2.3.1 `accessReporting` is a mandatory field while the corresponding `M5_ServiceAccessInformation.yaml` does not list it as required.

**Comment by davidjwbbc** [comment]
One thing I will point out is that the current *development* branch has `subprojects/open5gs` at commit 1ba7a73 (8th May 2023) and the *main* branch has it at f1c0b6d (28th Oct 2022). So it's possible the *main* branch is referencing the API v16 version of Open5GS and our code is now expecting the API v17 version.

Although I think both branches should now be on Open5GS v2.6.4 (commit 99f7da1)

**Comment by davidjwbbc** [comment]
The consumption reporting updates getting stuck looks like a definite bug to me, I think we should raise an issue for this and I'll look into it.

The discrepancy between TS26.512 Table 11.2.3.1 and the OpenAPI YAML should be fixed too, we should raise a Standards issue for this one. Although I'm unsure what to recommend as the there is also a discrepancy between which fields are mandatory between M1 and M5 APIs.

**Comment by davidjwbbc** [comment]
The "Assertion \`ogs_app()->sbi.client.key' failed" appears if you haven't updated your `msaf.yaml` configuration to include some new Open5GS fields: `sbi.client` and `sbi.server`. These are used to set client certificates for server side and client side HTTP communications.

I suggest the following change to your installed `msaf.yaml` (or delete and reinstall the latest default) :
```yaml
sbi:
    client:
        no_tls: true
    server:
        no_tls: true
```
This will disable the Open5GS SBI library from trying to look up keys and certificates.

**Reply by davidjwbbc** [reply]
The default Open5GS action for `ninja -C build install` or `meson install ...` is to not overwrite the configuration files, so you can end up with an out of date configuration file.
